File: soonwin-os-Python-Server/app/models/task.py
"""任务跟踪功能数据模型"""
from .. import db
from datetime import datetime


class Task(db.Model):
    """任务主表"""
    __tablename__ = 'task'

    id = db.Column(db.Integer, primary_key=True, autoincrement=True)
    author_id = db.Column(db.String(100), nullable=False, default='')
    author_name = db.Column(db.String(100), nullable=False, default='')
    content = db.Column(db.Text, nullable=False, default='')
    status = db.Column(db.String(20), nullable=False, default='pending')  # 'pending' | 'completed'
    completion_note = db.Column(db.Text, nullable=True)
    completion_image_url = db.Column(db.String(500), nullable=True)
    todo_image_url = db.Column(db.String(500), nullable=True)
    expected_date = db.Column(db.String(10), nullable=True)  # YYYY-MM-DD
    background_color = db.Column(db.String(20), nullable=True)  # hex 颜色
    like_count = db.Column(db.Integer, nullable=False, default=0)
    is_deleted = db.Column(db.Integer, nullable=False, default=0)
    created_at = db.Column(db.DateTime, nullable=False, default=datetime.now)
    updated_at = db.Column(db.DateTime, nullable=False, default=datetime.now, onupdate=datetime.now)
    completed_at = db.Column(db.DateTime, nullable=True)

    # 关系
    comments = db.relationship('TaskComment', backref='task', lazy='dynamic',
                                cascade='all, delete-orphan')
    visibilities = db.relationship('TaskVisibility', backref='task', lazy='dynamic',
                                    cascade='all, delete-orphan')
    likes = db.relationship('TaskLike', backref='task', lazy='dynamic',
                             cascade='all, delete-orphan')
    histories = db.relationship('TaskHistory', backref='task', lazy='dynamic',
                                 cascade='all, delete-orphan',
                                 order_by='TaskHistory.modified_at.desc()')

    def to_dict(self, include_relations=True):
        data = {
            'id': self.id,
            'author_id': self.author_id,
            'author_name': self.author_name,
            'content': self.content,
            'status': self.status,
            'completion_note': self.completion_note or '',
            'completion_image_url': self.completion_image_url or '',
            'todo_image_url': self.todo_image_url or '',
            'expected_date': self.expected_date or '',
            'background_color': self.background_color or '',
            'like_count': self.like_count or 0,
            'is_deleted': bool(self.is_deleted),
            'created_at': self.created_at.strftime('%Y-%m-%d %H:%M:%S') if self.created_at else None,
            'updated_at': self.updated_at.strftime('%Y-%m-%d %H:%M:%S') if self.updated_at else None,
            'completed_at': self.completed_at.strftime('%Y-%m-%d %H:%M:%S') if self.completed_at else None,
            'comment_count': self.comments.filter_by(is_deleted=0).count() if include_relations else 0,
            'history_count': self.histories.count() if include_relations else 0,
        }
        if include_relations:
            data['visibilities'] = [v.to_dict() for v in self.visibilities.all()]
        return data

File: soonwin-oa-VUE-FrontEnd/src/types/task.ts
/** 任务可见性配置 */
export interface TaskVisibility {
  id: number
  task_id: number
  visibility_type: 'role' | 'employee'
  visibility_value: string
  created_at: string
}

/** 任务 */
export interface Task {
  id: number
  author_id: string
  author_name: string
  content: string
  status: 'pending' | 'completed'
  completion_note: string
  completion_image_url: string
  todo_image_url: string
  expected_date: string  // YYYY-MM-DD 或 ''
  background_color: string  // hex 颜色，如 "#ff5500" 或 ''
  like_count: number
  is_deleted: boolean
  comment_count: number
  is_liked: boolean
  created_at: string
  updated_at: string
  completed_at: string | null
  visibilities?: TaskVisibility[]
}

/** 任务留言 */
export interface TaskComment {
  id: number
  task_id: number
  author_id: string
  author_name: string
  content: string
  is_deleted: boolean
  deleted_at: string | null
  deleted_by: string
  created_at: string
}

/** 任务修改历史 */
export interface TaskHistory {
  id: number
  task_id: number
  modified_by: string
  modified_at: string
  snapshot?: string
}

/** 任务列表响应 */
export interface TaskListResponse {
  tasks: Task[]
  total: number
  page: number
  per_page: number
  total_pages: number
}

/** 通知项（留言/点赞） */
export interface TaskNotificationComment {
  id: number
  task_id: number
  author_name: string
  author_id: string
  content: string
  created_at: string
  task_content_preview: string
}

export interface TaskNotificationLike {
  task_id: number
  user_id: string
  name: string
  created_at: string
  task_content_preview: string
}

export interface TaskNotificationsResponse {
  comments: TaskNotificationComment[]
  likes: TaskNotificationLike[]
  unread_count: number
}

/** 创建任务的表单 */
export interface TaskFormData {
  content: string
  expected_date?: string
  background_color?: string
  todo_image?: File
}